control/grpc.py

```diff
@@ -5982,65 +5978,27 @@ def list_subsystems_safe(self, request, context):
                 subsystem = pb2.subsystem_cli()
                 json_format.Parse(json.dumps(s), subsystem, ignore_unknown_fields=True)
                 subsystems.append(subsystem)
+                if not request.subsystem_nqn:
+                    cache_subsystem = pb2.subsystem()
+                    json_format.Parse(json.dumps(s), cache_subsystem, ignore_unknown_fields=True)
+                    cache_subsystems.append(cache_subsystem)
             except Exception:
                 self.logger.exception(f"{s=} parse error")
                 pass
 
+        # Only set cache if we've listed all subsystems
+        if not request.subsystem_nqn:
+            self.subsystems_cache.set_subsystems(pb2.subsystems_info(subsystems=cache_subsystems))
```

Comment:
When list_subsystems is invoked with a serial-number filter (request.serial_number set, subsystem_nqn unset), the loop above only appends the filtered results to cache_subsystems.